posts/progressive-enhancement/index.qmd

```diff
@@ -0,0 +1,152 @@
+
+In particular, this is the case when the old output was already inheriting from the parent of the new class (hence the importance of carefully choosing the parent class). In this situation, the new attributes from the new class should not interfere with existing code for downstream analysis.
+
+In this case, let's consider a function that was previously returning an unclassed `data.frame` with patient-level data:
+
+```{r}
+create_patient_dataset <- function(n_patients = 10) {
+
+  data <- data.frame(
+    patient_id = seq_len(n_patients),
+    age = sample(18:99, n_patients, replace = TRUE)
+  )
+
+  return(data)
+
+}
+```
+
+We want to start returning a `linelist` object. Because `linelist` objects are `data.frame`s (or `tibble`s) with an extra `attr`, it can be done in a transparent way:
+
+```{r}
+create_patient_dataset <- function(n_patients = 10) {
+
+  data <- data.frame(
+    patient_id = seq_len(n_patients),
+    age = sample(18:99, n_patients, replace = TRUE)
+  )
+
+  data <- linelist::make_linelist(
+    data,
+    id = "patient_id",
+    age = "age"
+  )
+
+  return(data)
+
+}
+
+inherits(data, "data.frame")
+```
+
+For a more realistic example, you can also see the work in progress to integrate the [new `contactmatrix` standard format](https://github.com/socialcontactdata/contactmatrix) for social contact data to the [contactdata package](https://github.com/Bisaloo/contactdata).
+
+This is however only true if code in downstream analysis follows good practices [^1]. If existing code was testing equality of the class to a certain value, it will break when the new class value is appended. This is described in a [post on the R developer blog, when base R was adding a new `array` class value to `matrix` objects](https://developer.r-project.org/Blog/public/2019/11/09/when-you-think-class.-think-again/index.html). Class inheritance should never be tested via `class(x) == "some_class"`. Instead, `inherits(x, "some_class")` or `is(x, "some_class")` should be used to future-proof the code and allow appending an additional in the future.
+
+[^1]: This is now [enforced in R packages by R CMD check](https://github.com/r-devel/r-svn/commit/77ebdff5adc200dfe9bc850bc4447088830d2ee0), and via the [`class_equals_linter()`](https://lintr.r-lib.org/reference/class_equals_linter.html)  in the [lintr package](https://lintr.r-lib.org/).
+
+## Conclusion
+
+Object oriented programming and S3 classes offer a convenient way to iteratively add interoperability in the ecosystem in a way that is minimally disruptive to users and developers.
+Newly classed input support can be added via custom methods (after converting the existing function to a generic if necessary).
+Newly classed output support can be added via progressive enhancement, by ensuring that the new class is a subclass of the old one, that downstream code uses good practices.
```

Comment:
The statement "that downstream code uses good practice" doesn't follow smoothly after the command but I'm unsure of what you're trying to say.